WIP BXMSDOC-8000 & BXMSDOC-8001: Update Kogito book as per 7.12 STP BAPLs (#72)
hmanwani-rh/kie-docs: doc-content/enterprise-only/getting-started-kogito/ref-kogito-remote-process-decision-traffic-example.adoc

[id="ref-kogito-remote-process-decision-traffic-example_{context}"]

> = Example traffic violations project for the remote method

Reviewer: Apply the comments that I made for the previous example to this example.

Author: Done

> The traffic violations project is used as an example for the remote method of integrating process and decision microservices in {KOGITO}.

Reviewer: ```suggestion
The traffic violations project is an example of the remote method of integrating process and decision microservices in {KOGITO}.
```

Author: Ditto

In the traffic violations project, first the information of the driver is fetched, and based on the fetched information, the driver’s license is validated using a RuleUnit in DRL. After the driver’s license is validated, the violation evaluation is executed that is defined as a DMN decision microservice. The output contains information about whether or not the driver is suspended.

The traffic violations example of the remote method contains the following microservices:

** DMN to define a decision microservice
** DRL to define rules decision microservice
** REST endpoints to integrate the process and decision microservices
** Service task to call an external decision microservice
** REST work item to call an external decision microservice

> IMPORTANT: The ID property of a process microservice is used in the REST endpoint generation, which refers to the path that is used to interact with the process microservice.

Reviewer: ```suggestion
IMPORTANT: The ID property of a process microservice is used to generate the REST endpoint, which refers to the path that is used to interact with the process microservice.
```

+
--
When using service tasks for defining a process microservice, you can provide an implementation in the application that executes the task, such as a REST or HTTP call.

.Traffic process microservice with service tasks
image::KogitoMicroservices/traffic-rules-dmn-service-task.png[Image of Traffic Process with service tasks for REST in process designer]

For implementation details, see the following `LicenseValidationRestClient` interface and `TrafficViolationRestClient` class.

.LicenseValidationRestClient class in traffic violations example
[source,java]
----
@Path("/validation/first")
@RegisterRestClient
public interface LicenseValidationRestClient {

    @POST
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    Driver post(Map<String, Object> parameters);

}
----

.TrafficViolationRestClient class in traffic violations example
[source,java]
----
@Path("/Traffic%20Violation")
@RegisterRestClient
public interface TrafficViolationRestClient {

    @POST
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    TrafficViolationResponse post(Map<String, Object> parameters);

}
----

> The process microservice using a service task is defined in the `traffic-rules-dmn-service-task.bpmn` file.

Reviewer 2: We could add here that in case of service tasks, an additional configuration in [application.properties](https://github.com/kiegroup/kogito-examples/blob/stable/process-decisions-rest-quarkus/src/main/resources/application.properties#L9-L11) is needed to "point" these 2 clients to the decision services. Remember, these services don't have to run in the same application or even on the same machine.

Author: Done

--

Traffic process microservice using REST work item::
+
--

> Using the REST work item is a declarative method, which does not require additional implementation and the REST or HTTP call is executed by the {DECISION_ENGINE}.

Reviewer 2: ```suggestion
Using the REST work item is a declarative method, which does not require additional implementation and the REST or HTTP call is executed by the {PROCESS_ENGINE}.
```

Author: Fixed it

.Process microservice using REST work item
image::KogitoMicroservices/traffic-rules-dmn-wih.png[Image of Traffic Process with service tasks for REST in process designer]

The information that is required to execute the REST call, such as URL and HTTP method must be defined in the *Data Assignments* property for the REST work item. The process microservice using REST work item is defined in the `traffic-rules-dmn-wih.bpmn` file.
--

In the traffic violations project, the `Violation` class is mapped with the `speedLimit` attribute using the `@JsonProperty` annotation. Therefore, the `speedLimit` attribute can be integrated with the `Violation` data type defined in the DMN model.

.`Violation` class defined in traffic process microservice
[source,java]
----
public class Violation {

    @JsonProperty("Code")
    private String code;

    @JsonProperty("Date")
    private Date date;

    @JsonProperty("Type")
    private String type;

    @JsonProperty("Speed Limit")
    private BigDecimal speedLimit;

    @JsonProperty("Actual Speed")
    private BigDecimal actualSpeed;

    //Getters / Setters
}
----

> In the traffic violations example, when fetching the driver information, an external call to a service task is performed to retrieve the driver information from the database. In this case, the service task implementation is performed using the `DriverService` class. In the *Data Assignments* property of the service task, the `driverId` is defined as an input variable and the `driver` is defined as an output variable, which contains the driver information.

Reviewer 2: ```suggestion
In the traffic violations example, when fetching the driver information, a mocked external call with a service task is performed to retrieve the driver information from the database. In this case, the service task implementation is performed using the `DriverService` class. In the *Data Assignments* property of the service task, the `driverId` is defined as an input variable and the `driver` is defined as an output variable, which contains the driver information.
```
It is the service task itself which issues the call and the call is mocked, see the [example](https://github.com/kiegroup/kogito-examples/blob/stable/process-decisions-rest-quarkus/src/main/java/org/kie/kogito/traffic/DriverService.java#L36-L40).

Author: Done
